llvm/lib/Transforms/Coroutines/CoroSplit.cpp

```diff
@@ -1967,22 +2047,13 @@ splitCoroutine(Function &F, SmallVectorImpl<Function *> &Clones,
   for (DbgVariableRecord *DVR : DbgVariableRecords)
     coro::salvageDebugInfo(ArgToAllocaMap, *DVR, Shape.OptimizeFrame,
                            false /*UseEntryValue*/);
-  return Shape;
-}
 
-/// Remove calls to llvm.coro.end in the original function.
-static void removeCoroEndsFromRampFunction(const coro::Shape &Shape) {
-  if (Shape.ABI != coro::ABI::Switch) {
-    for (auto *End : Shape.CoroEnds) {
-      replaceCoroEnd(End, Shape, Shape.FramePtr, /*in resume*/ false, nullptr);
-    }
-  } else {
-    for (llvm::AnyCoroEndInst *End : Shape.CoroEnds) {
-      auto &Context = End->getContext();
-      End->replaceAllUsesWith(ConstantInt::getFalse(Context));
-      End->eraseFromParent();
-    }
+  removeCoroEndsFromRampFunction(Shape);
+
+  if (!isNoSuspendCoroutine && Shape.ABI == coro::ABI::Switch) {
```

Comment:
Additionally, we should only create this `noalloc` variant, if the coroutines promise type itself is marked as "elidable".

**Example**

```
struct [[coro_must_elide]] ElidedTask { ... }
struct NonElidedTask { ... }

ElidedTask foo();
NonElidedTask bar();

NonElidedTask foobar() {
    co_await foo();
    co_await bar();
}
```

**Current behavior**: We create `foo.noalloc`, `bar.noalloc` and `foobar.alloc`.

`foo.noalloc` gets actually called. However, `bar.noalloc` and `foobar.alloc` are dead code. The `.noalloc` variants will never be used, as the `NonElidedTask` is not marked with `[[coro_must_elide]]`.

I think we should avoid generating this dead code

Reply:
We currently don't propagate the attribute from the CXXRecordDecl to the LLVM type. We can potentially add another attribute at the coroutine function level if that sounds good to you.

Reply:
Yes, I would be in favor of adding a second attribute. What do you & others think?
CC @ChuanqiXu9

Reply:
> Yes, I would be in favor of adding a second attribute. What do you & others think? CC @ChuanqiXu9

I can put up another PR for that. Wdyt?

Reply:
fine by me. Doesn't need to be addressed as part of this commit, can also be done in a follow-up commit.
Maybe add a `FIXME` to the source code, such that other people reading the source code in the meantime are aware of this future work?

Reply:
For sure!

Reply:
Agreed.

Reply:
This turned out to be easy. I am addressing this with the next push for this patch.